lib/notifier.ts

```diff
@@ -0,0 +1,40 @@
+import {ContractTransaction, ContractReceipt} from '@ethersproject/contracts';
+
+type Callback<T> = (tx: T) => void;
+
+export class Notifier<T> {
+  protected successCallbacks: Callback<T>[];
+  protected errorCallbacks: Callback<Error>[];
+
+  constructor(readonly promise: Promise<T>) {
+    this.successCallbacks = [];
+    this.errorCallbacks = [];
+    this._handlePromise(promise);
+  }
+
+  private _handlePromise(promise: Promise<T>): void {
+    promise
+      .then((res) => this.successCallbacks.forEach((f) => f(res)))
+      .catch((err) => this.errorCallbacks.forEach((f) => f(err)));
+  }
+
+  on<Error>(event: 'error', callback: Callback<Error>): void;
+  // eslint-disable-next-line no-dupe-class-members
+  on<T>(event: string, callback: Callback<T>): void;
+  // eslint-disable-next-line no-dupe-class-members
+  on(event: string, callback: any): void {
```

Comment:
Can we type the callback here?

Reply:
It might be doable but I couldn't get my away around the typescript type checker.
If I change this to `Callback<T> | Callback<Error>`, I am unsure how to get the correct type when calling the function. I guess `Callback<any>` would still be better than just `any`. If you know how to trick the type checker to work with the union type, let me know.